src/Microsoft.Diagnostics.Monitoring.RestServer/MetricsService.cs

```diff
@@ -20,53 +22,86 @@ internal sealed class MetricsService : BackgroundService
         private EventCounterPipeline _counterPipeline;
         private readonly IDiagnosticServices _services;
         private readonly MetricsStoreService _store;
-        private readonly MetricsOptions _options;
+        private IOptionsMonitor<MetricsOptions> _optionsMonitor;
 
         public MetricsService(IDiagnosticServices services,
-            IOptions<MetricsOptions> metricsOptions,
+            IOptionsMonitor<MetricsOptions> optionsMonitor,
             MetricsStoreService metricsStore)
         {
             _store = metricsStore;
             _services = services;
-            _options = metricsOptions.Value;
+            _optionsMonitor = optionsMonitor;
         }
         
-        protected override Task ExecuteAsync(CancellationToken stoppingToken)
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
         {
-            return Task.Run( async () =>
+            while (!stoppingToken.IsCancellationRequested)
             {
-                while (!stoppingToken.IsCancellationRequested)
+                stoppingToken.ThrowIfCancellationRequested();
+
+                try
                 {
-                    stoppingToken.ThrowIfCancellationRequested();
+                    //TODO In multi-process scenarios, how do we decide which process to choose?
+                    //One possibility is to enable metrics after a request to begin polling for metrics
+                    IProcessInfo pi = await _services.GetProcessAsync(filter: null, stoppingToken);
 
-                    try
-                    {
-                        //TODO In multi-process scenarios, how do we decide which process to choose?
-                        //One possibility is to enable metrics after a request to begin polling for metrics
-                        IProcessInfo pi = await _services.GetProcessAsync(filter: null, stoppingToken);
+                    var client = new DiagnosticsClient(pi.EndpointInfo.Endpoint);
 
-                        var client = new DiagnosticsClient(pi.EndpointInfo.Endpoint);
+                    MetricsOptions options = _optionsMonitor.CurrentValue;
+                    using var optionsTokenSource = new CancellationTokenSource();
 
-                        _counterPipeline = new EventCounterPipeline(client, new EventPipeCounterPipelineSettings
-                        {
-                            CounterGroups = Array.Empty<EventPipeCounterGroup>(),
-                            Duration = Timeout.InfiniteTimeSpan,
-                            RefreshInterval = TimeSpan.FromSeconds(_options.UpdateIntervalSeconds)
-                        }, metricsLogger: new[] { new MetricsLogger(_store.MetricsStore) });
+                    //If metric options change, we need to cancel the existing metrics pipeline and restart with the new settings.
```

Comment:
Should the existing metrics store be cleared when the options change?

Reply:
My initial thinking was that we can keep them, mainly because we timestamp them. Older metrics will still be there but it will be obvious that they have not been updated.